src/Data/Array/Polarized/Pull.hs

```diff
@@ -1,21 +1,85 @@
 {-# LANGUAGE LinearTypes #-}
 {-# LANGUAGE NoImplicitPrelude #-}
 
--- | This module is intended to be imported qualified, e.g.
+-- | This module provides pull arrays.
+--
+-- Use pull arrays to force fusion (i.e., force GHC to not allocate arrays in
+-- the garbage collector memory) in a computation that only uses lists
+-- as /intermediate/ results without operations that \"write\" to a list.
+-- You should think about pull arrays as arrays you can \"pull\" or read from,
+-- zip with and map over. These are arrays that work nicely as arguments to
+-- functions and poorly as arrays to write to.
+--
+-- Pull arrays fit within a larger framework for controlling when
+-- garbage collector memory is allocated; please see "Data.Array.Polarized".
+--
+-- Import this module qualified for clarity and to avoid
+-- name clashes.
+--
+-- == Example
+--
 -- > import qualified Data.Array.Polarized.Pull as Pull
+-- > import Data.Vector (Vector, (!), fromList)
+-- > import qualified Prelude as P
+-- >
+-- > type Pull a = Pull.Array a
+-- >
+-- > -- | compute the norm of summing three vectors
+-- > -- for vectors x,y,z, this computes || x + y + z ||
+-- > pullArrExample :: IO ()
+-- > pullArrExample = do
+-- >   x <- inputVectorX
+-- >   y <- inputVectorY
+-- >   z <- inputVectorZ
+-- >   let x_pull = Pull.fromVector x
+-- >   let y_pull = Pull.fromVector y
+-- >   let z_pull = Pull.fromVector z
+-- >   let nsum = normSumExample x_pull y_pull z_pull
+-- >   putStrLn P.$ "norm sum is " ++ show nsum
+-- >
+-- > normSumExample :: Pull Int -> Pull Int -> Pull Int -> Double
+-- > normSumExample x y z = norm (sumV (sumV x y) z)
+-- >
+-- > sumV :: Pull Int #-> Pull Int #-> Pull Int
+-- > sumV = Pull.zipWith (Linear.+)
+-- >
+-- > norm :: Pull Int -> Double
```

Comment:
You should really avoid function which require unrestricted access to pull or push arrays. They are not suited for this. It's not that it can't work (though, maybe we should make it harder to get such unrestricted access), but if you indeed share a pull array, you are duplicating computation, so it's a rather bad idea.

Example should not encourage bad practices.

Reply:
Yes, It doesn't need to be non-linear so I'll fix it.